ranger/data/scope.sh

```diff
@@ -290,6 +292,48 @@ handle_mime() {
             xls2csv -- "${FILE_PATH}" && exit 5
             exit 1;;
 
+        ## SQLite
+        *sqlite3)
+            ## Preview as text conversion
+            if [ -n "$( sqlite3 "file:${FILE_PATH}?mode=ro" '.tables' )" ]; then
+                sqlite_show_query() {
+                    sqlite-utils query "${FILE_PATH}" "${1}" --table --fmt fancy_grid ||
+                    sqlite3 "file:${FILE_PATH}?mode=ro" "${1}" -header -column
+                }
+                # Display basic table infomation
+                sqlite_rowcount_query="$( sqlite3 "file:${FILE_PATH}?mode=ro" -noheader \
+                    'SELECT (group_concat("SELECT """ || name || """ as tblname, count(*) as rowcount FROM " || name, " UNION ALL "))
+                    FROM sqlite_master WHERE type="table" AND name NOT LIKE "sqlite_%";' )"
+                sqlite_show_query \
+                    'SELECT tblname as "table", rowcount as "count",
+                    (
+                        SELECT "(" || group_concat(name, ", ") || ")"
+                        FROM pragma_table_info(tblname)
+                    ) as "columns",
+                    (
+                        SELECT "(" || group_concat(upper(type) || (CASE WHEN pk > 0 THEN " PRIMARY KEY" ELSE "" END), ", ") || ")"
+                        FROM pragma_table_info(tblname)
+                    ) as "types"
+                    FROM '"(${sqlite_rowcount_query});"
+                if [ "$SQLITE_TABLE_LIMIT" -gt 0 ] && [ "$SQLITE_ROW_LIMIT" -ge 0 ]; then
+                    # Do exhaustive preview
+                    echo; printf '>%.0s' $(seq "${PV_WIDTH}"); echo
+                    sqlite3 "file:${FILE_PATH}?mode=ro" -noheader \
+                        "SELECT name FROM sqlite_master WHERE type='table' AND name NOT LIKE 'sqlite_%' LIMIT ${SQLITE_TABLE_LIMIT};" |
+                        while read -r sqlite_table; do
+                            sqlite_table_rowcount="$(sqlite3 "file:${FILE_PATH}?mode=ro" -noheader "SELECT count(*) FROM ${sqlite_table}")"
+                            echo; echo "${sqlite_table}[${sqlite_table_rowcount}]:"
+                            sqlite_table_query="SELECT * FROM ${sqlite_table};"
+                            [ "$SQLITE_ROW_LIMIT" -gt 0 ] && sqlite_table_query="SELECT * FROM ${sqlite_table} LIMIT ${SQLITE_ROW_LIMIT} OFFSET (${sqlite_table_rowcount} - ${SQLITE_ROW_LIMIT});"
+                            sqlite_show_query "${sqlite_table_query}"
+                        done
+                fi
+                exit 5
+            elif [ -x "$( command -v sqlite3 )" ]; then
```

Comment:
The shell doesn't report a different return code when the command doesn't exist compared to what sqlite returns when the db is empty?

Reply:
`if [ -n "$( sqlite3 "file:${FILE_PATH}?mode=ro" '.tables' )" ]; then`
- `sqlite3` not installed: subcommand gets non-zero return code, condition fails.
- db is empty: subcommand return 0 with empty string, `[ -n "" ]`, condition fails.

Reply:
I replace `[ -x "$( command -v sqlite3 )" ]` with `sqlite3 "file:${FILE_PATH}?mode=ro" '.quit'`. No `[ -x $cmd ]` command check now.

Reply:
check command using `sqlite3 -version 1>/dev/null`